SSA/Experimental/Bits/Fast/BitStream.lean

```diff
@@ -380,11 +380,86 @@ local infix:20 " ≈ʷ " => EqualUpTo w
 @[simp] theorem ofBitVec_add : ofBitVec (x + y) ≈ʷ (ofBitVec x) + (ofBitVec y)  := by
   sorry
 
-@[simp] theorem ofBitVec_neg : ofBitVec (- x) ≈ʷ  - (ofBitVec x) := by
-  sorry
+/--
+Does the nth bit overflow/carry in calculating the negation of x?
+-/
+def doesNegCarry? (x : BitVec w) (n : Nat) : Bool := match n with
+  | Nat.zero => !x.getLsb 0
+  | Nat.succ y => !x.getLsb (Nat.succ y) &&  doesNegCarry? x y
+
+def iunfolder_state (x : BitVec w) (n : Nat) : Bool := match n with
+  | 0 => false
+  | i + 1 => doesNegCarry? x i
+
+lemma mod_mod (x : Nat) (w : Nat) (h : 0 < w) : x % 2 ^ w % 2 = x % 2 := by
+  -- We need to show that the least significant bit of x % 2 ^ w is the same as the least significant bit of x
+  -- Since 2 is a power of 2, the least significant bit of x % 2 ^ w is the same as the least significant bit of x
+  have y : 2 ^ 1 ∣ 2 ^ w := Nat.pow_dvd_pow 2 (by omega)
+  simp only [pow_one] at y
+  exact Nat.mod_mod_of_dvd x y
+
+theorem neg_mod_odd {w t : Nat} (hw : 0 < w) (h : t < 2 ^ w) : (2 ^ w - 1 - t + 1) % 2 = 1 ↔ t % 2 = 1 := by
```

Comment:
```suggestion
theorem twoPow_sub_mod_eq {w t : Nat} (hw : 0 < w) (h : t < 2 ^ w) : (2 ^ w - t) % 2 = t % 2 := by
```
Since `2 ^ w - 1 - t + 1 = 2 ^ w - t`, we should just use the latter, simpler, phrasing. 
Also, saying that `a = 1` iff `b = 1` is an odd way of saying that `a = b`